README.md

```diff
@@ -54,49 +67,161 @@ my_store = factory.get_hashstore(module_name, class_name, properties)
+object_cid = object_metadata.id
 
 # Store metadata (.../[hashstore_path]/metadata/)
 # By default, storing metadata will use the given properties namespace `format_id`
 pid = "j.tao.1700.1"
 sysmeta = "/path/to/your/sysmeta/document.xml"
 metadata_cid = my_store.store_metadata(pid, sysmeta)
-```
 
-If you want to store other types of metadata, add an additional `format_id`.
-```py
+# If you want to store other types of metadata, add an additional `format_id`.
 pid = "j.tao.1700.1"
 metadata = "/path/to/your/metadata/document.json"
 format_id = "http://custom.metadata.com/json/type/v1.0"
 metadata_cid = my_store.store_metadata(pid, metadata, format_id)
+
+# ...
+```
+
+###### Working with objects (store, retrieve, delete)
+
+In HashStore, objects are first saved as temporary files while their content identifiers are calculated. Once the default hash algorithm list and their hashes are generated, objects are stored in their permanent location using the store's algorithm's corresponding hash value, the store depth and the store width. Lastly, reference files are created for the object so that they can be found and retrieved given an identifier (ex. persistent identifier (pid)). Note: Objects are also stored once and only once.
+
+By calling the various interface methods for  `store_object`, the calling app/client can validate, store and tag an object simultaneously if the relevant data is available. In the absence of an identfiier (ex. persistent identifier (pid)), `store_object` can be called to solely store an object. The client is then expected to call `verify_object` when the relevant metadata is available to confirm that the object has been stored as expected. If the object is determined to be invalid (via `verify_object`), it will be deleted. Lastly, to finalize this process of storing an object (to make the object discoverable), the client calls `tag_object`. In summary, there are two expected paths to store an object:
+
+```py
+# All-in-one process which stores, validates and tags an object
+objectMetadata objInfo = store_object(InputStream, pid, additionalAlgorithm, checksum, checksumAlgorithm, objSize)
+
+# Manual Process
+# Store object
+obj_metadata = store_object(InputStream)
+# Validate object, throws exceptions if there is a mismatch and deletes the associated file
+verify_object(objInfo, checksum, checksumAlgorithn, objSize)
+# Tag object, makes the object discoverable (find, retrieve, delete)
+tag_object(pid, cid)
+```
+
+**How do I retrieve an object if I have the pid?**
+- To retrieve an object, call the Public API method `retrieve_object` which opens a stream to the object if it exists.
+
+**How do I find an object or check that it exists if I have the pid?**
+- To find the location of the object, call the Public API method `find_object` which will return the content identifier (cid) of the object.
+- This cid can then be used to locate the object on disk by following HashStore's store configuration.
```

Comment:
Initially, I wanted to have an "exists" method - which led to me using `find_object` for more than one purpose as I had documented above. In the worst case scenario where someone would have to go in and manually fix things, I thought this would be a helpful method... Only someone with the right privileges would be able to make use of this actual location. I would like to keep this in, but if you feel strongly that it shouldn't please let me know and let's discuss.

Reply:
Exposing the details of internal implementations to external clients is a fundamental design flaw, not just my preference :-)  https://en.wikipedia.org/wiki/Encapsulation_(computer_programming)

I have seen it lead to many serious situations IRL, where client code relies on these internal details, and then their code breaks when those details change.

If you think this method is needed, then it should not be be part of a public API, but should instead have some form of access restriction so it's usable only by admins, for example

Reply:
Note - this has been resolved (we have removed this from the Public API)